### PR TITLE
java: Add `FUNCTION FLUSH` command.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
https://redis.io/docs/latest/commands/function-flush/

Depends on #277 which depends on #270

### CODE REVIEW

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java [hunk 1]
```diff
@@ -485,7 +486,7 @@ private static Object[] scriptingAndFunctionsCommands(BaseTransaction<?> transac
                 };
 
         transaction
-                .customCommand(new String[] {"function", "flush", "sync"})
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/TransactionTestUtilities.java:488]: should we test functionFlush without the FlushMode argument or is it not necessary?

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 2]
```diff
@@ -291,4 +292,70 @@ public interface ScriptingAndFunctionsClusterCommands {
      * }</pre>
      */
     CompletableFuture<Map<String, Object>[]> functionListWithCode(String libNamePattern, Route route);
+
+    /**
+     * Deletes all the libraries.<br>
```
[the remark this answers is not in the document]
Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:297]: I remember we had similar discussion for `FLUSHALL` command (its behavior also depends on the same setting), but I can't find where.
Finally decided to keep javadocs short and not to include such details.

File: java/integTest/src/test/java/glide/TransactionTestUtilities.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/TransactionTestUtilities.java:488]: Added  to the end, thank you.

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (acarbonetto) [reply java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:297]: Let's be lazy and not do it.  

Honestly, users should be able to look up the Redis commands to get the latest and greatest documentation.  If we copy their documentation here, we may get in trouble with out of date documentation.

Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:297]:
```suggestion
     * Deletes all function libraries.<br>
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 3]
```diff
@@ -291,4 +292,70 @@ public interface ScriptingAndFunctionsClusterCommands {
...
+     * @example
+     *     <pre>{@code
+     * String response = client.functionFlush().get();
+     * assert response.equals("OK");
+     * }</pre>
+     */
+    CompletableFuture<String> functionFlush();
+
+    /**
+     * Deletes all the libraries.<br>
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:312]:
```suggestion
     * Deletes all function libraries.<br>
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 4]
```diff
@@ -291,4 +292,70 @@ public interface ScriptingAndFunctionsClusterCommands {
...
+     * @example
+     *     <pre>{@code
+     * String response = client.functionFlush(SYNC).get();
+     * assert response.equals("OK");
+     * }</pre>
+     */
+    CompletableFuture<String> functionFlush(FlushMode mode);
+
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:329]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java [hunk 5]
```diff
@@ -291,4 +292,70 @@ public interface ScriptingAndFunctionsClusterCommands {
...
+     * @example
+     *     <pre>{@code
+     * String response = client.functionFlush(RANDOM).get();
+     * assert response.equals("OK");
+     * }</pre>
+     */
+    CompletableFuture<String> functionFlush(Route route);
+
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsClusterCommands.java:345]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsCommands.java [hunk 6]
```diff
@@ -142,4 +143,34 @@ public interface ScriptingAndFunctionsCommands {
      * }</pre>
      */
     CompletableFuture<Map<String, Object>[]> functionListWithCode(String libNamePattern);
+
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsCommands.java:148]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsCommands.java [hunk 7]
```diff
@@ -142,4 +143,34 @@ public interface ScriptingAndFunctionsCommands {
...
+     * @example
+     *     <pre>{@code
+     * String response = client.functionFlush().get();
+     * assert response.equals("OK");
+     * }</pre>
+     */
+    CompletableFuture<String> functionFlush();
+
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/commands/ScriptingAndFunctionsCommands.java:162]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 8]
```diff
@@ -2982,6 +2983,32 @@ public T functionListWithCode(@NonNull String libNamePattern) {
         return getThis();
     }
 
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2987]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 9]
```diff
@@ -2982,6 +2983,32 @@ public T functionListWithCode(@NonNull String libNamePattern) {
...
+     * @see <a href="https://redis.io/docs/latest/commands/function-flush/">redis.io</a> for details.
+     * @return Command Response - <code>OK</code>.
+     */
+    public T functionFlush() {
+        protobufTransaction.addCommands(buildCommand(FunctionFlush));
+        return getThis();
+    }
+
+    /**
+     * Deletes all the libraries.
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2999]:
```suggestion
     * Deletes all function libraries.
```

File: java/client/src/main/java/glide/api/models/commands/FlushMode.java [hunk 10]
```diff
@@ -4,17 +4,27 @@
...
 import glide.api.RedisClusterClient;
 import glide.api.models.configuration.RequestRoutingConfiguration.Route;
+import glide.api.models.configuration.RequestRoutingConfiguration.SingleNodeRoute;
 
+// TODO add links to script flush
 /**
- * Defines flushing mode for <code>FLUSHALL</code> command implemented by {@link
- * RedisClient#flushall(FlushMode)}, {@link RedisClusterClient#flushall(FlushMode)}, and {@link
- * RedisClusterClient#flushall(FlushMode, Route)}.
+ * Defines flushing mode for
```
Comment (acarbonetto) [inline java/client/src/main/java/glide/api/models/commands/FlushMode.java:11]:
```suggestion
 * Defines flushing mode for:
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 11]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
...
     }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_returns_success() {
+        // setup
+        String[] args = new String[] {};
+        String value = "42";
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4286]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 12]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
         assertEquals(testResponse, response);
         assertEquals(value, payload);
     }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_returns_success() {
+        // setup
+        String[] args = new String[] {};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4284]: remove

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 13]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
...
+        when(commandManager.<String>submitNewCommand(eq(FunctionFlush), eq(args), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<String> response = service.functionFlush();
+        String payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertEquals(value, payload);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4298]:
```suggestion
        assertEquals(OK, payload);
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 14]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
...
+    }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_with_mode_returns_success() {
+        // setup
+        FlushMode mode = ASYNC;
+        String[] args = new String[] {mode.toString()};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4307]: remove

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 15]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
...
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4309]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/test/java/glide/api/RedisClientTest.java [hunk 16]
```diff
@@ -4272,4 +4275,49 @@ public void functionListWithCode_with_pattern_returns_success() {
...
+        when(commandManager.<String>submitNewCommand(eq(FunctionFlush), eq(args), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<String> response = service.functionFlush(mode);
+        String payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertEquals(value, payload);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClientTest.java:4321]:
```suggestion
        assertEquals(OK, payload);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 17]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
         assertEquals(testResponse, response);
         assertEquals(value, payload);
     }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_returns_success() {
+        // setup
+        String[] args = new String[] {};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1377]: remove

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 18]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1379]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 19]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        assertEquals(value, payload);
+    }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_with_mode_returns_success() {
+        // setup
+        FlushMode mode = ASYNC;
+        String[] args = new String[] {mode.toString()};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1400]: remove

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 20]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1402]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 21]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        when(commandManager.<String>submitNewCommand(eq(FunctionFlush), eq(args), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<String> response = service.functionFlush(mode);
+        String payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertEquals(value, payload);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1414]:
```suggestion
        assertEquals(OK, payload);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 22]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+    }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_with_route_returns_success() {
+        // setup
+        String[] args = new String[] {};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1422]: remove

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 23]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1424]:
```suggestion
        testResponse.complete(OK);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 24]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        assertEquals(value, payload);
+    }
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_with_mode_and_route_returns_success() {
+        // setup
+        FlushMode mode = ASYNC;
+        String[] args = new String[] {mode.toString()};
+        String value = "42";
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1445]: remove

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 25]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+        when(commandManager.<String>submitNewCommand(eq(FunctionFlush), eq(args), eq(RANDOM), any()))
+                .thenReturn(testResponse);
+
+        // exercise
+        CompletableFuture<String> response = service.functionFlush(mode, RANDOM);
+        String payload = response.get();
+
+        // verify
+        assertEquals(testResponse, response);
+        assertEquals(value, payload);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1459]:
```suggestion
        assertEquals(OK, payload);
```

File: java/client/src/test/java/glide/api/RedisClusterClientTest.java [hunk 26]
```diff
@@ -1365,4 +1368,94 @@ public void functionListWithCode_with_pattern_and_route_returns_success() {
...
+
+    @SneakyThrows
+    @Test
+    public void functionFlush_with_mode_and_route_returns_success() {
+        // setup
+        FlushMode mode = ASYNC;
+        String[] args = new String[] {mode.toString()};
+        String value = "42";
+        CompletableFuture<String> testResponse = new CompletableFuture<>();
+        testResponse.complete(value);
```
Comment (acarbonetto) [inline java/client/src/test/java/glide/api/RedisClusterClientTest.java:1447]:
```suggestion
        testResponse.complete(OK);
```

**Review by acarbonetto — APPROVED** [review]

**Review by aaron-congo — APPROVED** [review]

**Pull request merged**